File: swift_mort/MORT/Models/Application.swift
import Foundation

struct MortApplication: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let jobID: UUID
    let teenID: UUID
    let status: String
    let note: String?
    let guardianID: UUID?
    let job: Job?
    let availabilityConfirmed: Bool
    let createdAt: String?
    let viewedAt: String?
    let withdrawnAt: String?
    let teen: ProfileSummary?

    enum CodingKeys: String, CodingKey {
        case id, status, note, teen
        case jobID = "job_id"
        case teenID = "teen_id"
        case guardianID = "guardian_id"
        case job = "jobs"
        case availabilityConfirmed = "availability_confirmed"
        case createdAt = "created_at"
        case viewedAt = "viewed_at"
        case withdrawnAt = "withdrawn_at"
    }
}

struct ApplicationEligibility: Codable, Hashable, Sendable {
    let eligible: Bool
    let code: String
    let message: String
    let guardianRequiredForThisJob: Bool
    let guardianLinked: Bool
    let verificationRequirement: String
    let scheduleType: String

    enum CodingKeys: String, CodingKey {
        case eligible, code, message
        case guardianRequiredForThisJob = "guardian_required_for_this_job"
        case guardianLinked = "guardian_linked"
        case verificationRequirement = "verification_requirement"
        case scheduleType = "schedule_type"
    }
}

struct ApplicationStatusEvent: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let applicationID: UUID
    let fromStatus: String?
    let toStatus: String
    let actorID: UUID?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case applicationID = "application_id"
        case fromStatus = "from_status"
        case toStatus = "to_status"
        case actorID = "actor_id"
        case createdAt = "created_at"
    }
}

File: flutter_mort/lib/data/models/application.dart
import 'job.dart';

class MortApplication {
  const MortApplication({
    required this.id,
    required this.jobId,
    required this.teenId,
    required this.status,
    this.note,
    this.guardianId,
    this.job,
    this.availabilityConfirmed = false,
    this.createdAt,
    this.viewedAt,
    this.withdrawnAt,
    this.teenName,
    this.teenAvatarPath,
  });

  final String id;
  final String jobId;
  final String teenId;
  final String status;
  final String? note;
  final String? guardianId;
  final Job? job;
  final bool availabilityConfirmed;
  final DateTime? createdAt;
  final DateTime? viewedAt;
  final DateTime? withdrawnAt;
  final String? teenName;
  final String? teenAvatarPath;

  factory MortApplication.fromMap(Map<String, dynamic> json) {
    final jobData = json['jobs'];
    final teenData = json['teen'];
    final teen = teenData is Map<String, dynamic> ? teenData : null;
    return MortApplication(
      id: json['id'].toString(),
      jobId: json['job_id'].toString(),
      teenId: json['teen_id'].toString(),
      status: (json['status'] as String?) ?? 'submitted',
      note: json['note'] as String?,
      guardianId: json['guardian_id'] as String?,
      job: jobData is Map<String, dynamic> ? Job.fromMap(jobData) : null,
      availabilityConfirmed: json['availability_confirmed'] == true,
      createdAt: DateTime.tryParse((json['created_at'] ?? '').toString()),
      viewedAt: DateTime.tryParse((json['viewed_at'] ?? '').toString()),
      withdrawnAt: DateTime.tryParse((json['withdrawn_at'] ?? '').toString()),
      teenName: teen?['display_name'] as String?,
      teenAvatarPath: teen?['avatar_path'] as String?,
    );
  }
}

class ApplicationEligibility {
  const ApplicationEligibility({
    required this.eligible,
    required this.code,
    required this.message,
    required this.guardianRequiredForThisJob,
    required this.guardianLinked,
    this.verificationRequirement = 'none',
    this.scheduleType = 'flexible',
  });

  final bool eligible;
  final String code;
  final String message;
  final bool guardianRequiredForThisJob;
  final bool guardianLinked;
  final String verificationRequirement;
  final String scheduleType;

  factory ApplicationEligibility.fromMap(Map<String, dynamic> map) {
    return ApplicationEligibility(
      eligible: map['eligible'] == true,
      code: (map['code'] as String?) ?? 'unknown_permission_failure',
      message: (map['message'] as String?) ?? 'We could not check eligibility.',
      guardianRequiredForThisJob: map['guardian_required_for_this_job'] == true,
      guardianLinked: map['guardian_linked'] == true,
      verificationRequirement:
          (map['verification_requirement'] as String?) ?? 'none',
      scheduleType: (map['schedule_type'] as String?) ?? 'flexible',
    );
  }
}
